README.md
# Moment timezone for local time

Get all matching timezones given a local time and a timestamp.  
Moment-timezone-for-local-time is a Moment.js plugin.

```js
// Given a timestamp, for which timezones is the local time 03:00 ?
moment.tz.forLocalTime(3, 0, new Date("2000-01-01T00:00:00Z"));
// -> ["Africa/Algiers", ..., "Europe/Paris", ...]
```

## Installation

In browser:

```html
<script src="https://unpkg.com/moment-timezone-for-local-time"></script>
<!-- Plugin is then available at window.momentTimezoneForLocalTime -->
```

In Node:

```bash
npm install moment-timezone-for-local-time
```

```js
import * as momentTzForLocalTime from "moment-timezone-for-local-time"; // ES Modules
const momentTzForLocalTime = require("moment-timezone-for-local-time"); // CommonJS
```

You may need to install moment-timezone if not already installed.

## Version compatibility

Indicates the version to use according your moment-timezone version.

| moment-timezone | moment-timezone-for-local-time | notes                                              |
| --------------- | ------------------------------ | -------------------------------------------------- |
| >= 0.5.30       | 1.1.0                          |
| >= 0.5.29       | 1.0.0                          | use at least moment 2.15.0 to avoid type conflicts |
| >= 0.2.0        | 0.1.1                          |
| 0.1.0           | 0.1.0                          |

## Setup

Let's extend the moment instance

```js
import * as moment from "moment-timezone";
momentTzForLocalTime.extend(moment); // Adds moment.tz.forLocalTime function
moment.tz.forLocalTime();
```

It's also possible to not extend the moment instance

```js
const forLocalTime = momentTzForLocalTime.forLocalTime.bind(moment.tz);
forLocalTime();
```

## Usage

```js
// fn(hour, minute, timestamp)
forLocalTime(3, 0, new Date("2000-01-01T00:00:00Z"));
forLocalTime(3, 0, new Date("2000-01-01T00:00:00Z").getTime());
forLocalTime(3, 0, moment("2000-01-01T00:00:00Z"));
// The above 3 calls are equal

// timestamp defaults to Date.now()
forLocalTime(10, 30);

// List all timezones with the same local time as current system
forLocalTime();
```

Returns an array of timezone identifiers e.g. `["Africa/Algiers", ..., "Europe/Paris", ...]`.  
The returned array can be empty when local time and timestamp don't match any timezone.

A `RangeError` will be thrown when:

- hour isn't in \[0, 23\]
- minute isn't in \[0, 59\]

Decimals will be floored.

```js
// The below calls are equal
forLocalTime(3.9, 0.9, 0.9);
forLocalTime(3, 0, 0);
```

The result of calls to `forLocalTime` depends on timezone data loaded into moment. Make sure you load the right dataset for your use (e.g. 10 year range, 1970-2030, ...).

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## Dev

Fullfill the peer dependency if npm doesn't install it for you.

```bash
npm install --no-save moment-timezone@version
```

for `version` refer to package.json#peerDependencies.moment-timezone.

## License

[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
# 1.1.0

- Types update. `namespace tz` becomes `interface MomentTimezone`. That reflects the moment-timezone v0.5.30 definition file.

# 1.0.0

**BREAKING CHANGES:**
The package now use the in-built types of moment-timezone. The type of arguments is narrowed.

- Type of `this` for `forLocaltime` was `any` and is now `typeof moment.tz`.
- Type `extend` now requires a `typeof moment`.

# 0.1.1

- Update tests according to tzdata of moment-timezone 0.2.0.
- (No actual changes otherwise)

# 0.1.0

## Features

- `extend(moment)` method which add the function `forLocalTime` to `moment.tz`.
- `forLocalTime` can be used as a standalone without mutating `moment`.


test/data.ts
export interface DataCtx {
  date: Date;
  isoString: string;
  localtime: {
    hour: number;
    minute: number;
  };
  timestamp: number;
  timezones: string[];
}

export function newYear2000UTC1(): DataCtx {
  const isoString = "2000-01-01T00:00:00Z";
  const date = new Date(isoString);
  const localtime = { hour: 1, minute: 0 };
  /** in ms */
  const timestamp = 946684800000;
  // For each following timezone, localtime for 2000-01-01T00:00:00Z is 01:00
  // Daylight Saving Time rules:
  // - EU, W-Eur, E-Eur, C-Eur -> Mar last Sun, Oct last Sun
  // - WAT no DST, UTC+1 constant
  const timezones = [
    "Africa/Algiers", // EU
    "Africa/Bangui", // WAT
    "Africa/Brazzaville", // WAT
    "Africa/Ceuta", // EU
    "Africa/Douala", // WAT
    "Africa/Kinshasa", // WAT
    "Africa/Lagos", // WAT
    "Africa/Libreville", // WAT
    "Africa/Luanda", // WAT
    "Africa/Malabo", // WAT
    "Africa/Ndjamena", // WAT
    "Africa/Niamey", // WAT
    "Africa/Porto-Novo", // WAT
    "Africa/Tunis", // EU
    "Arctic/Longyearbyen", // EU
    "Atlantic/Jan_Mayen", // EU
    "CET", // EU
    "Etc/GMT-1", // none
    "Europe/Amsterdam", // EU
    "Europe/Andorra", // EU
    "Europe/Belgrade", // EU
    "Europe/Berlin", // EU
    "Europe/Bratislava", // EU
    "Europe/Brussels", // EU
    "Europe/Budapest", // EU
    "Europe/Busingen", // EU
    "Europe/Copenhagen", // EU
    "Europe/Gibraltar", // EU
    "Europe/Ljubljana", // EU
    "Europe/Luxembourg", // EU
    "Europe/Madrid", // EU
    "Europe/Malta", // EU
    "Europe/Monaco", // EU
    "Europe/Oslo", // EU
    "Europe/Paris", // EU
    "Europe/Podgorica", // EU
    "Europe/Prague", // EU
    "Europe/Rome", // EU
    "Europe/San_Marino", // EU
    "Europe/Sarajevo", // EU
    "Europe/Skopje", // EU
    "Europe/Stockholm", // EU
    "Europe/Tirane", // EU
    "Europe/Vaduz", // EU
    "Europe/Vatican", // EU
    "Europe/Vienna", // EU
    "Europe/Warsaw", // EU
    "Europe/Zagreb", // EU
    "Europe/Zurich", // EU
    "MET", // none
    "Poland", // EU
  ];
  return { date, timezones, localtime, timestamp, isoString };
}


test/index.test.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import * as moment from "moment-timezone";
import { extend, forLocalTime } from "../src/index";
import { forLocalTime as _forLocalTime } from "../src/moment-timezone-for-local-time";
jest.mock("../src/moment-timezone-for-local-time");
import { DataCtx, newYear2000UTC1 } from "./data";

extend(moment);

it("should extend throw if given object not in shape of moment", () => {
  expect(() => extend(null as any)).toThrow(
    "The given object doesn't have a 'tz' property. Make sure to import moment-timezone before extending momentjs with moment-timezone-for-local-time."
  );
});

it("should extend modify moment so that exports.forLocalTime === moment.tz.forLocalTime", () => {
  // eslint-disable-next-line @typescript-eslint/unbound-method
  expect(moment.tz.forLocalTime).toBe(forLocalTime);
});

describe("moment.tz.fromLocalTime:", () => {
  // It's easier to mock Date instead of setting the node timezone
  // on Unix it's easy but on Windows we need to change system's timezone

  let data: DataCtx;
  let now: Date;
  let dateOG: DateConstructor;

  beforeAll(() => {
    data = newYear2000UTC1();
    dateOG = global.Date;
    global.Date = (class extends Date {
      public static now() {
        return data.timestamp;
      }
      constructor() {
        super(data.isoString);
      }
      public getHours() {
        return data.localtime.hour;
      }
      public getMinutes() {
        return data.localtime.minute;
      }
      public getTime() {
        return data.timestamp;
      }
      public valueOf() {
        return data.timestamp;
      }
    } as unknown) as DateConstructor;
  });

  beforeEach(() => {
    now = new Date();
  });

  it("when no args given, should call the internal with current system local time", () => {
    moment.tz.forLocalTime();
    expect(_forLocalTime).toHaveBeenCalledTimes(1);
    expect(_forLocalTime).toHaveBeenCalledWith(
      now.getHours(),
      now.getMinutes(),
      now.getTime()
    );
  });

  describe("timestamp should defaults to current system time when argument is", () => {
    it("null", () => {
      moment.tz.forLocalTime(5, 0, null);
      expect(_forLocalTime).toHaveBeenCalledTimes(1);
      expect(_forLocalTime).toHaveBeenCalledWith(5, 0, now.getTime());
    });

    it("undefined", () => {
      moment.tz.forLocalTime(5, 0, undefined);
      expect(_forLocalTime).toHaveBeenCalledTimes(1);
      expect(_forLocalTime).toHaveBeenCalledWith(5, 0, now.getTime());
    });
  });

  it("timestamp can be either be a Date, number or moment and should the return value should be the same whatever the argument type", () => {
    const curried = (timestamp: Parameters<typeof forLocalTime>[2]): string[] =>
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      moment.tz.forLocalTime(now.getHours(), now.getMinutes(), timestamp);
    const expected = curried(now.getTime()); // number
    expect(curried(moment(data.timestamp))).toEqual(expected); // moment instance
    expect(curried(now)).toEqual(expected); // Date
  });

  afterAll(() => {
    global.Date = dateOG;
  });
});

describe("moment.tz.fromLocalTime: should throw when", () => {
  it("hour is not a number", () => {
    expect(() => {
      moment.tz.forLocalTime("foo" as any, 0);
    }).toThrow("hour arg must be a number and in [0,23]. You gave foo.");
    expect(() => {
      moment.tz.forLocalTime(null as any, 0);
    }).toThrow(
      "hour arg must be a number and in [0,23]. You gave null or undefined."
    );
  });
  it("hour not in valid range (> 23)", () => {
    expect(() => {
      moment.tz.forLocalTime(24, 0);
    }).toThrow("hour arg must be a number and in [0,23]. You gave 24.");
  });
  it("hour not in valid range (< 0)", () => {
    expect(() => {
      moment.tz.forLocalTime(-1, 0);
    }).toThrow("hour arg must be a number and in [0,23]. You gave -1.");
  });
  it("minute is not a number", () => {
    expect(() => {
      moment.tz.forLocalTime(0, "foo" as any);
    }).toThrow("minute arg must be a number and in [0,59]. You gave foo.");
    expect(() => {
      moment.tz.forLocalTime(0, null as any);
    }).toThrow(
      "minute arg must be a number and in [0,59]. You gave null or undefined."
    );
  });
  it("minute not in valid range (> 23)", () => {
    expect(() => {
      moment.tz.forLocalTime(0, 60);
    }).toThrow("minute arg must be a number and in [0,59]. You gave 60.");
  });
  it("minute not in valid range (< 0)", () => {
    expect(() => {
      moment.tz.forLocalTime(0, -1);
    }).toThrow("minute arg must be a number and in [0,59]. You gave -1.");
  });
});


test/moment-timezone-for-local-time.test.ts
import * as moment from "moment-timezone";
import { forLocalTime as _forLocalTime } from "../src/moment-timezone-for-local-time";
import { newYear2000UTC1 } from "./data";

type OmitThisArg<F> = F extends (this: never, ...args: infer P) => infer R
  ? (...args: P) => R
  : never;

const forLocalTime = _forLocalTime.bind(moment.tz) as OmitThisArg<
  typeof _forLocalTime
>;

it("distant from DST transition", () => {
  const data = newYear2000UTC1();
  expect(
    forLocalTime(
      data.localtime.hour,
      data.localtime.minute,
      data.date.getTime()
    )
  ).toEqual(data.timezones);
});

/*
  We're going to use Europe/Paris
  IANA tzdb 2021a:

  # Zone  NAME          STDOFF  RULES	  FORMAT	[UNTIL]
    Zone  Europe/Paris  1:00    France  CE%sT	  1977
                        1:00    EU      CE%sT

  # Rule  NAME  FROM  TO   -  IN   ON       AT     SAVE  LETTER/S
    Rule  EU    1981  max  -  Mar  lastSun  1:00u  1:00  S
    Rule  EU    1996  max  -  Oct  lastSun  1:00u  0	-
*/

describe("Forward DST transition", () => {
  it("before transition", () => {
    const timestamp = new Date("2000-03-26T01:00:00Z").getTime() - 1;
    const res = forLocalTime(1, 59, timestamp);
    expect(res).toContain("Europe/Paris");
  });

  describe("should not contain the timezone when localtime is invalid (inside transition gap)", () => {
    it("at start", () => {
      const ref = new Date("2000-03-26T01:00:00Z");
      const res = forLocalTime(2, 0, ref.getTime());
      expect(res).not.toContain("Europe/Paris");
    });

    it("inside", () => {
      const ref = new Date("2000-03-26T01:00:00Z");
      const res = forLocalTime(2, 30, ref.getTime());
      expect(res).not.toContain("Europe/Paris");
    });
  });

  it("wall clock after transition", () => {
    const ref = new Date("2000-03-26T01:00:00Z");
    const res = forLocalTime(3, 0, ref.getTime());
    expect(res).toContain("Europe/Paris");
  });
});

it("should return empty array when localtime doesn't match with any timezone", () => {
  expect(
    forLocalTime(23, 59, new Date("2000-01-01T00:00:00Z").getTime())
  ).toEqual([]);
});

describe("Fallback DST transition", () => {
  it("overlap start", () => {
    const ref = new Date("2000-10-29T00:00:00Z");
    const res = forLocalTime(2, 0, ref.getTime());
    expect(res).toContain("Europe/Paris");
  });

  it("inside overlap", () => {
    const ref = new Date("2000-10-29T00:30:00Z");
    const res = forLocalTime(2, 30, ref.getTime());
    expect(res).toContain("Europe/Paris");
  });

  it("overlap end", () => {
    const ref = new Date("2000-10-29T01:00:00Z");
    const res = forLocalTime(2, 0, ref.getTime());
    expect(res).toContain("Europe/Paris");
  });

  it("invalid", () => {
    const ref = new Date("2000-10-29T01:00:00Z");
    const res = forLocalTime(3, 0, ref.getTime());
    expect(res).not.toContain("Europe/Paris");
  });
});

describe("should floor decimals", () => {
  it("hour arg", () => {
    const data = newYear2000UTC1();
    expect(
      forLocalTime(
        data.localtime.hour + 0.9,
        data.localtime.minute,
        data.date.getTime()
      )
    ).toEqual(
      forLocalTime(
        data.localtime.hour,
        data.localtime.minute,
        data.date.getTime()
      )
    );
  });

  it("minute arg", () => {
    const data = newYear2000UTC1();
    expect(
      forLocalTime(
        data.localtime.hour,
        data.localtime.minute + 0.9,
        data.date.getTime()
      )
    ).toEqual(
      forLocalTime(
        data.localtime.hour,
        data.localtime.minute,
        data.date.getTime()
      )
    );
  });

  it("timestamp arg", () => {
    const data = newYear2000UTC1();
    expect(
      forLocalTime(
        data.localtime.hour + 0.9,
        data.localtime.minute,
        data.date.getTime() + 0.9
      )
    ).toEqual(
      forLocalTime(
        data.localtime.hour,
        data.localtime.minute,
        data.date.getTime()
      )
    );
  });
});


src/index.ts
import moment from "moment-timezone";
import { forLocalTime as _forLocalTime } from "./moment-timezone-for-local-time";

declare module "moment" {
  interface MomentTimezone {
    /**
     * List all timezones with the same local time as current system.
     *
     * @returns list of timezone identifiers
     * @author [moment-timezone-for-local-time](https://github.com/Serrulien/moment-timezone-for-local-time)
     */
    forLocalTime(): string[];
    /**
     * List all timezones with the same local time for a given timestamp.
     *
     * @param timestamp defaults to Date.now()
     * @returns list of timezone identifiers
     * @throws {RangeError} when hour isn't in [0, 23] or minute isn't in [0, 59]
     * @author [moment-timezone-for-local-time](https://github.com/Serrulien/moment-timezone-for-local-time)
     */
    forLocalTime(
      hour: number,
      minute: number,
      timestamp?: moment.Moment | Date | number | null
    ): string[];
  }
}

/**
 * List all timezones with the same local time as current system.
 *
 * @returns list of timezone identifiers
 * @author [moment-timezone-for-local-time](https://github.com/Serrulien/moment-timezone-for-local-time)
 */
export function forLocalTime(this: typeof moment.tz): string[];
/**
 * List all timezones with the same local time for a given timestamp.
 *
 * @param timestamp defaults to Date.now()
 * @returns list of timezone identifiers
 * @throws {RangeError} when hour isn't in [0, 23] or minute isn't in [0, 59]
 * @author [moment-timezone-for-local-time](https://github.com/Serrulien/moment-timezone-for-local-time)
 */
export function forLocalTime(
  this: typeof moment.tz,
  hour: number,
  minute: number,
  timestamp?: moment.Moment | Date | number | null
): string[];
export function forLocalTime(
  this: typeof moment.tz,
  hour?: number,
  minute?: number,
  timestamp?: moment.Moment | Date | number | null
): string[] {
  if (arguments.length === 0) {
    const now = new Date();
    return _forLocalTime.call(
      this,
      now.getHours(),
      now.getMinutes(),
      now.getTime()
    );
  }
  if (timestamp === undefined || timestamp === null) {
    timestamp = Date.now();
  }
  if (typeof hour !== "number" || 23 < hour || hour < 0) {
    throw new TypeError(
      `hour arg must be a number and in [0,23]. You gave ${
        hour?.toString() || "null or undefined"
      }.`
    );
  }
  if (typeof minute !== "number" || 59 < minute || minute < 0) {
    throw new TypeError(
      `minute arg must be a number and in [0,59]. You gave ${
        minute?.toString() || "null or undefined"
      }.`
    );
  }
  return _forLocalTime.call(this, hour, minute, timestamp.valueOf());
}

export function extend(momentJS: typeof moment): void {
  if (!momentJS?.tz) {
    throw new Error(
      "The given object doesn't have a 'tz' property. Make sure to import moment-timezone before extending momentjs with moment-timezone-for-local-time."
    );
  }
  momentJS.tz.forLocalTime = forLocalTime;
}
